nft-rfc-006.md

```diff
@@ -0,0 +1,860 @@
+features for interacting with the subject. It may also provides IID
+specific properties:
+
+**AssociatedResources** -- provides a privacy-enabled way to attach
+digital resources associated with the asset. An optional array of one or
+more resource descriptors, it provides the metadata required for
+accessing and using that resource, e.g., type of resource, a proof to
+verify the resource, and a service endpoint for requesting that
+resource.
+
+Resources may be provided in-line or by secure reference. Inline
+resources allow appropriate use cases to directly include the resource
+itself in the IID Document. In many cases, this is a privacy problem.
+However, for some use cases, resources must be specified for
+on-chain execution, justifying the added bytes and potential disclosure risk.
+The resource descriptor provides for a flexible representation of
+various mime types, compression, and encoding, as befitting the use.
+
+This approach allows token owners to manage privacy in three key ways:
+
+1.  Avoids posting potentially sensitive information on-chain and hence,
+    in an unavoidably public and irrevocable manner.
+
+2.  Provides for a service endpoint that can apply appropriate privacy
+    and security checks before revealing information.
+
+3.  The hash graph resource descriptor type obscures not only the
+    content of the associated resource, but also the quantity.
+
+Resources may be secured by specifying a proofType of hash or hashgraph.
+A hashgraph uses a merkle tree of hashes for external content associated
+with this asset. A resource descriptor of this type obscures both the
+type and the number of such resources, while allowing each such resource
+to be verifiably associated. It also provides for privacy-respecting
+verification of complete disclosure. Anyone who needs to prove they have
+all of the associated resources can compare their own hash graph of
+resources with that stored in the IID Document. Note that this
+anti-censorship technique requires a verifier to discover the type and
+nature of those resources on their own.
+
+Proposed properties for resource descriptors in the AssociatedResources
+property:
+```json
+{ 
+    "AssociatedResources" : [{
+
+    "path": fully qualified IID Resource ID for this resource, e.g.,
+did:example:abc/myResource.png,
+
+    "type" : "nft:ResourceDescriptor",
+
+    "proofType" (optional) : "hash" | "hashgraph",
+
+    "proof" (optional): hash | hashgraph,
+
+    "resourceFormat": "IID Resource type | mime type",
+
+    "encoding" : "native" | "multibase" \| "string",
+
+    "compression": gzip \| none,
+
+    "encryption" : \[open ended for arbitrary extensibility\],
+
+    "identifier" (optional) : a unique reference for the resource in the
+asset's namespace,
+
+    "endpoint" (optional): url,
+
+    "resource" (optional): a representation of the resource for inline
+distribution
+
+}]
+}
+```
+
+### Types of Associated Resources for IIDs
```

Comment:
What are the differences between `Associated Resources` and `Services`?

Reply:
That's a great question.

I think that services *could* be used in the fashion we are describing for Associated Resources, with a few differences. And that idea deserves more consideration.

The simplest distinction is that is not expected that service endpoints themselves either embed the resource or provide a simple merkle hash. Nor is it expected that service endpoints point to different resources--they are understood to be ways of interacting with the DID Subject.

This is, in part, because the mental model is different. Services are expected to be interactive, while resources are expected to be static. Of course, the web shows that this distinction is a mirage, but outside of the NFT work, I've not seen service endpoints used to embed arbitrary assets nor refer to off-chain assets by hash.

There has been some effort to add a "resource" property to DID Core--and I believe some methods do intend to embed *the* resource identified by the DID within the DID Document. This is problematic--and I argued successfully to get it removed from DID Core because it puts in Core a feature that would distinguish human Subjects from digital resources, which violates herd privacy, which posits that one should not be able to discern the nature of the Subject from the DID or DID Document itself, except that which might be implied by the DID Method--because Methods are free to do whatever they want.

DIDs refer to a single subject. As such, the DID Document's properties are semantically interpreted as being about that subject. That is, if a DID refers to a schema, then the properties in the DID Document are about that schema--which is the use case where embedded distribution was advocated. Similarly, service endpoints are expected to be for interacting **with the Subject**.

IIDs on the other hand, **always** have the asset as the Subject. So, service endpoints would be more appropriately used for ways that you can interact with the on-chain asset.

Associated resources and references are NOT the on-chain asset but rather they are related resources and references. As such, its a simpler mental model to have any number of associated resources--they aren't *THE* resource identified by the IID, they are something else. As such, it is also much more amenable from a privacy perspective to have multiple resources and references. Yes, doing so does enable correlation between those resources, but the hashgraph resource types allows blinding of the actual resource to avoid such correlation, without precluding disclosure of those correlatable resources when appropriate. 

Finally, the ability to embed resources, link to resources, and cryptographically identify resources, all in the same structure, treats each type of association as a first class option. None is preferred or expected over the other, much as CSS elements in HTML, style attributes on html elements, and externally linked CSS files all operate equivalently when rendering HTML. An AssociatedResources doesn't anticipate embedded resources any more than it implies that the resource *is* the subject of the IID--both of which were problematic with DIDs because they violate herd privacy. Different IIDs can ensure different privacy protections. And anyone who supports IIDs will treat them all interoperably.

I'm intrigued by your suggestion because it does suggest that we could possibly define a service type that allows an embedded resource. However, after writing this explanation up, I think the biggest shift is that separation of Subject from resources. Service endpoints are for interacting with the DID Subject (which for IIDs is the on-chain asset). Associated Resources is for verifiably identifying (and potentially retrieving) resources other than the IID Subject. That level of indirection is probably a good reason NOT to use service endpoints for associated resources.